## Librerías y Paquetes

In [1]:
import pandas as pd 
import numpy as np

pd.options.display.float_format = '{:,.2f}'.format

## 1. Carga de los dataset

In [9]:
credit_df = pd.read_csv("../data/credit_record.csv")
credit_df

ID  MONTHS_BALANCE STATUS
0        5001711               0      X
1        5001711              -1      0
2        5001711              -2      0
3        5001711              -3      0
4        5001712               0      C
...          ...             ...    ...
1048570  5150487             -25      C
1048571  5150487             -26      C
1048572  5150487             -27      C
1048573  5150487             -28      C
1048574  5150487             -29      C

[1048575 rows x 3 columns]

In [3]:
application_df = pd.read_csv('../data/application_record.csv')
application_df

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008805           M            Y               Y             0   
2       5008806           M            Y               Y             0   
3       5008808           F            N               Y             0   
4       5008809           F            N               Y             0   
...         ...         ...          ...             ...           ...   
438552  6840104           M            N               Y             0   
438553  6840222           F            N               N             0   
438554  6841878           F            N               N             0   
438555  6842765           F            N               Y             0   
438556  6842885           F            N               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0             427,500.00               Working               Higher education   
1             427,500.00               Working               Higher education   
2             112,500.00               Working  Secondary / secondary special   
3             270,000.00  Commercial associate  Secondary / secondary special   
4             270,000.00  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
438552        135,000.00             Pensioner  Secondary / secondary special   
438553        103,500.00               Working  Secondary / secondary special   
438554         54,000.00  Commercial associate               Higher education   
438555         72,000.00             Pensioner  Secondary / secondary special   
438556        121,500.00               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0             Civil marriage   Rented apartment      -12005          -4542   
1             Civil marriage   Rented apartment      -12005          -4542   
2                    Married  House / apartment      -21474          -1134   
3       Single / not married  House / apartment      -19110          -3051   
4       Single / not married  House / apartment      -19110          -3051   
...                      ...                ...         ...            ...   
438552             Separated  House / apartment      -22717         365243   
438553  Single / not married  House / apartment      -15939          -3007   
438554  Single / not married       With parents       -8169           -372   
438555               Married  House / apartment      -21673         365243   
438556               Married  House / apartment      -18858          -1201   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                0           0           0  Security staff   
3                1                0           1           1     Sales staff   
4                1                0           1           1     Sales staff   
...            ...              ...         ...         ...             ...   
438552           1                0           0           0             NaN   
438553           1                0           0           0        Laborers   
438554           1                1           0           0     Sales staff   
438555           1                0           0           0             NaN   
438556           1                0           1           0     Sales staff   

        CNT_FAM_MEMBERS  
0                  2.00  
1                  2.00  
2                  2.00  
3                  1.00  
4                  1.00  
...                 ...  
438552             1.00  
438553             1.00  
438554             1.00  
43855

## 2. Join de los datos



In [10]:
credit_df.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [11]:
# Actualizar el valor de la columna status por valores numéricos
credit_df['STATUS'] = credit_df['STATUS'].replace({'X':-2, 'C':-1})
credit_df['STATUS'] = credit_df['STATUS'].astype('int')
credit_df.head()

ID  MONTHS_BALANCE  STATUS
0  5001711               0      -2
1  5001711              -1       0
2  5001711              -2       0
3  5001711              -3       0
4  5001712               0      -1

In [14]:
# Encontramos el último mes
credit_last_month = credit_df.groupby('ID', as_index=False).agg({'MONTHS_BALANCE': 'max'})
credit_last_month

ID  MONTHS_BALANCE
0      5001711               0
1      5001712               0
2      5001713               0
3      5001714               0
4      5001715               0
...        ...             ...
45980  5150482             -11
45981  5150483               0
45982  5150484               0
45983  5150485               0
45984  5150487               0

[45985 rows x 2 columns]

In [16]:
# Join para obtener el status del último mes
credit_record = pd.merge(credit_last_month, credit_df, how= 'inner', on= ['ID','MONTHS_BALANCE'] )
credit_record

ID  MONTHS_BALANCE  STATUS
0      5001711               0      -2
1      5001712               0      -1
2      5001713               0      -2
3      5001714               0      -2
4      5001715               0      -2
...        ...             ...     ...
45980  5150482             -11      -1
45981  5150483               0      -2
45982  5150484               0      -1
45983  5150485               0       0
45984  5150487               0      -1

[45985 rows x 3 columns]

In [17]:
# Join entre aplication y creadit para crear el df final
df = pd.merge(application_df, credit_record, how= 'inner', on= ['ID'])
df

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0      5008804           M            Y               Y             0   
1      5008805           M            Y               Y             0   
2      5008806           M            Y               Y             0   
3      5008808           F            N               Y             0   
4      5008809           F            N               Y             0   
...        ...         ...          ...             ...           ...   
36452  5149828           M            Y               Y             0   
36453  5149834           F            N               Y             0   
36454  5149838           F            N               Y             0   
36455  5150049           F            N               Y             0   
36456  5150337           M            N               Y             0   

       AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0            427,500.00               Working               Higher education   
1            427,500.00               Working               Higher education   
2            112,500.00               Working  Secondary / secondary special   
3            270,000.00  Commercial associate  Secondary / secondary special   
4            270,000.00  Commercial associate  Secondary / secondary special   
...                 ...                   ...                            ...   
36452        315,000.00               Working  Secondary / secondary special   
36453        157,500.00  Commercial associate               Higher education   
36454        157,500.00             Pensioner               Higher education   
36455        283,500.00               Working  Secondary / secondary special   
36456        112,500.00               Working  Secondary / secondary special   

         NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0            Civil marriage   Rented apartment      -12005          -4542   
1            Civil marriage   Rented apartment      -12005          -4542   
2                   Married  House / apartment      -21474          -1134   
3      Single / not married  House / apartment      -19110          -3051   
4      Single / not married  House / apartment      -19110          -3051   
...                     ...                ...         ...            ...   
36452               Married  House / apartment      -17348          -2420   
36453               Married  House / apartment      -12387          -1325   
36454               Married  House / apartment      -12387          -1325   
36455               Married  House / apartment      -17958           -655   
36456  Single / not married   Rented apartment       -9188          -1193   

       FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0               1                1           0           0             NaN   
1               1                1           0           0             NaN   
2               1                0           0           0  Security staff   
3               1                0           1           1     Sales staff   
4               1                0           1           1     Sales staff   
...           ...              ...         ...         ...             ...   
36452           1                0           0           0        Managers   
36453           1                0           1           1  Medicine staff   
36454           1                0           1           1  Medicine staff   
36455           1                0           0           0     Sales staff   
36456           1                0           0           0        Laborers   

       CNT_FAM_MEMBERS  MONTHS_BALANCE  STATUS  
0                 2.00               0      -1  
1                 2.00               0      -1  
2                 2.00               0      -1  
3                 1.00               0       0  
4                 1.00             -22      -2  
...               

## 2. Análisis de los datasets

In [7]:
credit_df.describe()

ID  MONTHS_BALANCE
count 1,048,575.00    1,048,575.00
mean  5,068,286.42          -19.14
std      46,150.58           14.02
min   5,001,711.00          -60.00
25%   5,023,644.00          -29.00
50%   5,062,104.00          -17.00
75%   5,113,856.00           -7.00
max   5,150,487.00            0.00

In [8]:
# Validamos el número de ID de clientes
credit_df['ID'].value_counts()

5145767    61
5116236    61
5037153    61
5002523    61
5028973    61
           ..
5149044     1
5079047     1
5137258     1
5023604     1
5023029     1
Name: ID, Length: 45985, dtype: int64

### 2.1. Application dataset

In [47]:
len(application_df)

438557

In [48]:
application_df.isna().sum()

ID                     0
CODE_GENDER            0
FLAG_OWN_CAR           0
FLAG_OWN_REALTY        0
CNT_CHILDREN           0
AMT_INCOME_TOTAL       0
NAME_INCOME_TYPE       0
NAME_EDUCATION_TYPE    0
NAME_FAMILY_STATUS     0
NAME_HOUSING_TYPE      0
DAYS_BIRTH             0
DAYS_EMPLOYED          0
FLAG_MOBIL             0
FLAG_WORK_PHONE        0
FLAG_PHONE             0
FLAG_EMAIL             0
CNT_FAM_MEMBERS        0
dtype: int64

El dataset *application_df* cuenta con un total de 438557 observaciones. Únicamente la variable OCCUPATION_TYPE cuenta con un gran número de valores nulos, equivalente al 30.6% del total de registros. Por lo que esta variable no será tomada en cuenta para el presente proyecto, ya que el procentaje de valores nulos es muy alto para rellenarlo utilizando algún método. 

In [28]:
application_df = application_df.drop(columns=['OCCUPATION_TYPE'])

In [29]:
# Análisis del tipo de datos
application_df.dtypes

ID                       int64
CODE_GENDER             object
FLAG_OWN_CAR            object
FLAG_OWN_REALTY         object
CNT_CHILDREN             int64
AMT_INCOME_TOTAL       float64
NAME_INCOME_TYPE        object
NAME_EDUCATION_TYPE     object
NAME_FAMILY_STATUS      object
NAME_HOUSING_TYPE       object
DAYS_BIRTH               int64
DAYS_EMPLOYED            int64
FLAG_MOBIL               int64
FLAG_WORK_PHONE          int64
FLAG_PHONE               int64
FLAG_EMAIL               int64
CNT_FAM_MEMBERS        float64
dtype: object

In [50]:
# Análisis de variables numéricas
application_df[['AMT_INCOME_TOTAL','CNT_FAM_MEMBERS','CNT_CHILDREN','DAYS_EMPLOYED','DAYS_BIRTH']].describe()

AMT_INCOME_TOTAL  CNT_FAM_MEMBERS  CNT_CHILDREN  DAYS_EMPLOYED  \
count        438,557.00       438,557.00    438,557.00     438,557.00   
mean         187,524.29             2.19          0.43      60,563.68   
std          110,086.85             0.90          0.72     138,767.80   
min           26,100.00             1.00          0.00     -17,531.00   
25%          121,500.00             2.00          0.00      -3,103.00   
50%          160,780.50             2.00          0.00      -1,467.00   
75%          225,000.00             3.00          1.00        -371.00   
max        6,750,000.00            20.00         19.00     365,243.00   

       DAYS_BIRTH  
count  438,557.00  
mean   -15,997.90  
std      4,185.03  
min    -25,201.00  
25%    -19,483.00  
50%    -15,630.00  
75%    -12,514.00  
max     -7,489.00

In [53]:
# Análisis de variable categóricas
for column in ['CODE_GENDER','FLAG_OWN_CAR','FLAG_OWN_REALTY','NAME_INCOME_TYPE','NAME_EDUCATION_TYPE','NAME_FAMILY_STATUS','NAME_HOUSING_TYPE','FLAG_MOBIL','FLAG_WORK_PHONE','FLAG_PHONE','FLAG_EMAIL']:
    print('--------------------')
    print(application_df[column].value_counts())

--------------------
F    294440
M    144117
Name: CODE_GENDER, dtype: int64
--------------------
N    275459
Y    163098
Name: FLAG_OWN_CAR, dtype: int64
--------------------
Y    304074
N    134483
Name: FLAG_OWN_REALTY, dtype: int64
--------------------
Working                 226104
Commercial associate    100757
Pensioner                75493
State servant            36186
Student                     17
Name: NAME_INCOME_TYPE, dtype: int64
--------------------
Secondary / secondary special    301821
Higher education                 117522
Incomplete higher                 14851
Lower secondary                    4051
Academic degree                     312
Name: NAME_EDUCATION_TYPE, dtype: int64
--------------------
Married                 299828
Single / not married     55271
Civil marriage           36532
Separated                27251
Widow                    19675
Name: NAME_FAMILY_STATUS, dtype: int64
--------------------
House / apartment      393831
With parents            